### ISSUE TITLE
M2.3.2 - Node <payment><payflow_advanced><user> is declared twice in Paypal module so it still use MCrypt

### ISSUE THREAD

**Issue by Linek** [magento/magento2#23880]
### Preconditions
1. Magento 2.3.2 Open Source & Enterprise

### Steps to reproduce
1. Flush cache
2. Add breakpoint in \Magento\Framework\Encryption\Encryptor::decrypt in line 437 (can be var_dump or die method). Just to make sure it won't go inside, as this is not supported method
3. Run the homepage

### Expected result
1. Website should run correctly without going inside of unsupported method

### Actual result
1. Website is stopping (die) on breakpoint because unsupported crypt is used

The reason is probably that in config.php for Magento_Paypal we have <user> node twice under <payflow_advanced> node.

```
<payflow_advanced>
        <model>Magento\Paypal\Model\Payflowadvanced</model>
        <payment_action>Authorization</payment_action>
        <verbosity>HIGH</verbosity>
            <user backend_model="Magento\Config\Model\Config\Backend\Encrypted" />
        <pwd backend_model="Magento\Config\Model\Config\Backend\Encrypted" />
        <group>paypal</group>
        <title>Credit Card</title>
        <partner>PayPal</partner>
        <vendor>PayPal</vendor>
            <user>PayPal</user>
        <csc_required>1</csc_required>
        <csc_editable>1</csc_editable>
        <url_method>GET</url_method>
        <email_confirmation>0</email_confirmation>
        <verify_peer>1</verify_peer>
        <transaction_url_test_mode>https://pilot-payflowpro.paypal.com</transaction_url_test_mode>
        <transaction_url>https://payflowpro.paypal.com</transaction_url>
        <cgi_url_test_mode>https://pilot-payflowlink.paypal.com</cgi_url_test_mode>
        <cgi_url>https://payflowlink.paypal.com</cgi_url>
</payflow_advanced>
```

And it tries to decrypt "PayPal" value.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-18656` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by Yupik** [comment]
#mm19pl

**Comment by Yupik** [comment]
Hello @Linek - as far as i can see, method \Magento\Framework\Encryption\Encryptor::decrypt() on branch 2.3-develop is fully supported (there's not any deprecated content). It's true, that Magento uses this method after clearing cache, but there's nothing wrong (it decrypt's username and password stored in database).

Right now Magento should go inside this method until PayPal module is active. Please correct me if i'm wrong here. Additional details could be helpful too - until providing i'll close this issue

**Comment by Linek** [comment]
Hi @Yupik 

Maybe I was not fully clear on my first description but in my opinion problem still exists.

So if you will review \Magento\Framework\Encryption\Encryptor::decrypt method there is this part:

```
           } elseif (1 === $partsCount) {
                $keyVersion = 0;
                $cryptVersion = self::CIPHER_BLOWFISH;
                // not supported format
            } else {
```

Normally the execution should never go into this condition. But because of configuration in config.xml from Magento_Paypal module, there is an execution of \Magento\Framework\Encryption\Encryptor::decrypt with String parameter "PayPal".

In my opinion it's wrong and there shouldn't be double xml node <user> under <payflow_advanced> node, but if in your opinion it's correct, then we can close it, but please mind that because of this the default BlackFire profile for Magento Cloud is displaying error that mcrypt is used while it shouldn't be anymore in 2.3

**Comment by Yupik** [comment]
Hello @Linek,

What do you exacly mean by:

> double xml node under <payflow_advanced> node

I don't see any double xml node under `<payflow_advanced>` node. I've tried to debug this, and on fresh Magento 2.3 install it goes to `decrypt()` method with empty string, not `PayPal` (maybe you've provided this string to username or password field?).

**Comment by Linek** [comment]
Hi @Yupik 

Could you please verify that your <payflow_advanced> looks like this:

```
            <payflow_advanced>
                <model>Magento\Paypal\Model\Payflowadvanced</model>
                <payment_action>Authorization</payment_action>
                <verbosity>HIGH</verbosity>
                <user backend_model="Magento\Config\Model\Config\Backend\Encrypted" />
                <pwd backend_model="Magento\Config\Model\Config\Backend\Encrypted" />
                <group>paypal</group>
                <title>Credit Card</title>
                <partner>PayPal</partner>
                <vendor>PayPal</vendor>
                <user>PayPal</user>
                <csc_required>1</csc_required>
                <csc_editable>1</csc_editable>
                <url_method>GET</url_method>
                <email_confirmation>0</email_confirmation>
                <verify_peer>1</verify_peer>
                <transaction_url_test_mode>https://pilot-payflowpro.paypal.com</transaction_url_test_mode>
                <transaction_url>https://payflowpro.paypal.com</transaction_url>
                <cgi_url_test_mode>https://pilot-payflowlink.paypal.com</cgi_url_test_mode>
                <cgi_url>https://payflowlink.paypal.com</cgi_url>
            </payflow_advanced>
```

So the <user> node is first defined with backend_model and then by string "PayPal".

BR

**Comment by Yupik** [comment]
@Linek Yes, it looks exacly like this, but still don't get this: 

> the node is first defined with backend_model and then by string "PayPal".

If you're talking about:
```<user backend_model="Magento\Config\Model\Config\Backend\Encrypted" />```
```<pwd backend_model="Magento\Config\Model\Config\Backend\Encrypted" />```

It's not defined twice. `backend_model` determines class which will handle value under `user` and `pwd` fields. It can be added to additional fields here, but we wan't to encrypt only username and password.

Could you please explain "So the node is first defined with backend_model and then by string "PayPal"." with code examples from this xml file?

**Comment by hostep** [comment]
@Yupik, just read the title of the issue, the `<user>` node is defined twice 😉 

<img width="665" alt="Screenshot 2019-09-19 at 13 24 40" src="https://user-images.githubusercontent.com/85479/65241778-bbd1d380-dae4-11e9-823a-aa45635f7adb.png">

- https://github.com/magento/magento2/blob/3ee05a4c4f2a6fb229126f3f3036562543fa318f/app/code/Magento/Paypal/etc/config.xml#L161
- https://github.com/magento/magento2/blob/3ee05a4c4f2a6fb229126f3f3036562543fa318f/app/code/Magento/Paypal/etc/config.xml#L167

**Comment by Yupik** [comment]
@hostep Ah, that's right, thought in different way about this title :) So i'll take care about this, but after fixing doubled node `<user>` and filling PayPal secret's - it still will go through `decrypt` method

**Comment by Linek** [comment]
Hi @Yupik 

There is nothing wrong with going through `decrypt` method. The problem is when parameter is of `decrypt` is "PayPal" because in this case it will use fallback and run Mcrypt method instead of Sodium.

**Comment by magento-engcom-team** [comment]
Hi @Linek. Thank you for your report. 
The issue has been fixed in magento/magento2#24694 by @Yupik in 2.3-develop branch
Related commit(s):
 - [998cd74f78b23d32dfab8456300382779454f99a](https://github.com/magento/magento2/commit/998cd74f78b23d32dfab8456300382779454f99a)
 - [f1e225d81bbb4dbb2b155531445cfc3e880db81d](https://github.com/magento/magento2/commit/f1e225d81bbb4dbb2b155531445cfc3e880db81d)

The fix will be available with the upcoming 2.3.4 release.
